File: app/src/main/java/com/lapptelier/test/TestActivity.java
package com.lapptelier.test;

import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;

import com.lapptelier.smartrecyclerview.DrawableDividerItemDecoration;
import com.lapptelier.smartrecyclerview.MultiGenericAdapter;
import com.lapptelier.smartrecyclerview.SmartRecyclerView;
import com.lapptelier.smartrecyclerview.ViewHolderInteractionListener;
import com.socks.library.KLog;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author L'Apptelier SARL
 * @date 14/09/2017
 */
public class TestActivity extends AppCompatActivity implements ViewHolderInteractionListener {

    @BindView(R.id.test_smart_recycler_view)
    SmartRecyclerView mRecyclerView;

    MultiGenericAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_test);
        ButterKnife.bind(this);

        final List<String> elements = Arrays.asList("test 1", "test 2", "test 3", "test 4", "test 5", "test 6", "test 7", "test 8");

        //configuration de la liste
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        mRecyclerView.setRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                KLog.d("FRESH-FRESH-FRESH !");
                mRecyclerView.displaySwipeToRefresh(false);
            }
        });

        // Configuration de l'adapter
        adapter = new MultiGenericAdapter(String.class, TestViewHolder.class, R.layout.cell_test, this);
        mRecyclerView.setAdapter(adapter);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            mRecyclerView.addItemDecoration(new DrawableDividerItemDecoration(getDrawable(R.drawable.divider), null, true));
        } else {
            mRecyclerView.addItemDecoration(new DrawableDividerItemDecoration(getResources().getDrawable(R.drawable.divider), null, true));
        }

        //on sette le texte de la vue vide
        mRecyclerView.setLoadingLayout(R.layout.empty);

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                adapter.addAll(elements);
            }
        }, 500);


    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    @Override
    public void onItemClicked(@NotNull Object item) {
        adapter.removeAt(adapter.getObjectIndex(item));
    }

    @Override
    public void onItemAction(@NotNull Object item, int action) {

    }
}

File: app/src/main/java/com/lapptelier/test/TestActivity.kt
package com.lapptelier.test

import android.os.Bundle
import android.os.Handler
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.snackbar.Snackbar
import com.lapptelier.smartrecyclerview.*
import com.lapptelier.test.databinding.ActivityTestBinding

/**
 * @author L'Apptelier SARL
 * @date 14/09/2017
 */
internal class TestActivity : AppCompatActivity(), ViewHolderInteractionListener, OnMoreListener,
    SwipeRefreshLayout.OnRefreshListener {

    private lateinit var adapter: TestAdapter
    private lateinit var binding: ActivityTestBinding

    private val elements = listOf(
        "test 1",
        "test 2",
        "test 3",
        "test 4",
        "test 5",
        "test 6",
        "test 7",
        "test 8",
        "test 9",
        "test 10",
        "test 11",
        "test 12",
        "test 13",
        "test 14",
        "test 15",
        "test 16"
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityTestBinding.inflate(layoutInflater)
        setContentView(binding.root)

        //configuration de la liste
        binding.testSmartRecyclerView.setLayoutManager(
            LinearLayoutManager(
                this,
                LinearLayoutManager.VERTICAL,
                false
            )
        )

        // Configuration de l'adapter
        adapter = TestAdapter(this)
        binding.testSmartRecyclerView.setAdapter(adapter, true)
        binding.testSmartRecyclerView.addItemDecoration(
            DrawableDividerItemDecoration(
                ContextCompat.getDrawable(
                    this,
                    R.drawable.divider
                ), null, true
            )
        )
        binding.testSmartRecyclerView.emptyLayout = R.layout.empty
        binding.testSmartRecyclerView.loadingLayout = R.layout.loading
        binding.testSmartRecyclerView.setOnMoreListener(this, 2)
        binding.testSmartRecyclerView.setRefreshListener(this)

        binding.testSmartRecyclerView.displayLoadingView()

        adapter.hasMore = true
        adapter.addAll(elements)

    }


    // OnMoreListener
    override fun onMoreAsked(
        overallItemsCount: Int,
        itemsBeforeMore: Int,
        maxLastVisiblePosition: Int
    ) {
        Log.d("test", "MORE MORE MORE !")

        Handler().postDelayed({
            adapter.hasMore = true
        }, 500)

        Handler().postDelayed({
            adapter.appendAll(elements)
        }, 1000)

    }

    override fun onRefresh() {
        Log.d("test", "FRESH-FRESH-FRESH !")
        Handler().postDelayed({
            adapter.hasMore = true
        }, 1000)

        Handler().postDelayed({
            adapter.addAll(elements)
        }, 2000)

    }


    override fun onItemAction(
        item: Any,
        viewId: Int,
        viewHolderInteraction: ViewHolderInteraction
    ) {
        if (item is String)
            when (viewHolderInteraction) {
                ViewHolderInteraction.LONG_TAP -> adapter.removeAt(adapter.getObjectIndex(item))
                else -> Snackbar.make(
                    binding.root,
                    "Tapped :$item",
                    Snackbar.LENGTH_LONG
                ).show()
            }
    }
}
